main/Main.hs

```diff
@@ -2,180 +2,290 @@
-    H.statement "SELECT current_setting('server_version_num')::integer"
-      HE.unit (HD.singleRow $ HD.value HD.int4) False
-
 {-|
+  The purpose of this worker is to fill the refDbStructure created in 'main'
+  with the 'DbStructure' returned from calling 'getDbStructure'. This method
+  is meant to be called by multiple times by the same thead, but does nothing if 
+  the previous invocation has not terminated. In all cases this method does not 
+  halt the calling thead, the work is proformed in a separate thread.
+
+  Note: 'atomicWriteIORef' is essentialy a lazy semaphore that prevents two 
+  threads from running 'connectionWorker' at the same time. 
+
   Background thread that does the following :
   1. Tries to connect to pg server and will keep trying until success.
-  2. Checks if the pg version is supported and if it's not it kills the main program.
+  2. Checks if the pg version is supported and if it's not it kills the main 
+     program.
   3. Obtains the dbStructure.
-  4. If 2 or 3 fail to give their result it means the connection is down so it goes back to 1,
-     otherwise it finishes his work successfully.
+  4. If 2 or 3 fail to give their result it means the connection is down so it 
+     goes back to 1, otherwise it finishes his work successfully.
 -}
-connectionWorker :: ThreadId -> P.Pool -> Schema -> IORef (Maybe DbStructure) -> IORef Bool -> IO ()
+connectionWorker
+  :: ThreadId -- main Thread's Id
+  -> P.Pool   -- Connection Pool to PostgreSQL
+  -> Schema   -- Schema PostgREST generates the API off of
+  -> IORef (Maybe DbStructure) -- The pointer that the worker will fill in
+  -> IORef Bool -- Binary Semaphore (must be accessed using atomicWriteIORef)
+  -> IO ()
 connectionWorker mainTid pool schema refDbStructure refIsWorkerOn = do
   isWorkerOn <- readIORef refIsWorkerOn
   unless isWorkerOn $ do
     atomicWriteIORef refIsWorkerOn True
     void $ forkIO work
   where
+    pgVersion =
+      H.statement
+        "SELECT current_setting('server_version_num')::integer"
+        HE.unit
+        (HD.singleRow $ HD.value HD.int4)
+        False
     work = do
       atomicWriteIORef refDbStructure Nothing
       putStrLn ("Attempting to connect to the database..." :: Text)
-      connected <- connectingSucceeded pool
+      connected <- connectingSucceeded
       when connected $ do
-        result <- P.use pool $ do
-          supported <- isServerVersionSupported
-          unless supported $ liftIO $ do
-            hPutStrLn stderr
-              ("Cannot run in this PostgreSQL version, PostgREST needs at least "
-              <> pgvName minimumPgVersion)
-            killThread mainTid
-          dbStructure <- getDbStructure schema
-          liftIO $ atomicWriteIORef refDbStructure $ Just dbStructure
+        result <-
+          P.use pool $ do
+            supported <-
+              fmap
+                (\ver -> ver >= pgvNum minimumPgVersion)
+                (H.query () pgVersion)
```

Comment:
The ```isServerVersionSupported``` was more concise here(though maybe it could be renamed to ```isPgVersionSupported```). Previous ```connectionWorker``` was already a long function, by merging the 3 functions it got longer, the indentation increased and there are nested where's, overall I think this is less clear than before. Maybe it would be better to go back to having these as top-level functions.